### ISSUE TITLE
Proposal for "VK V1 Incremental Evolution"

### ISSUE THREAD

**Issue by sargun** [virtual-kubelet/virtual-kubelet#964]
**VK 1-NG**

This is a proposal on how we can evolve the internal VK state reconciliation mechanism. The goal is to remain API compatible to 1.0, and not evolve where we do not need to.

One of the things that’s grown more complicated over the time in the VK is the management of state “Edges” -- If you imagine the VK as managing a bunch of state machines, and each state is a vertex, we take change events from the Provider or API Server and try to reconcile them.

I suggest instead we shift to a new approach, where instead, all pods are reconciled with more Kubernetes-esque line-level logic. Since the VK knows all pods that are scheduled on it, we can scan over these.

The pod would look something like this (details non-specific):

```go
type podstate struct {
  Lock sync.Mutex 
  podReceivedFromAPIServer *v1.Pod
  podReceivedFromProvider *v1.Pod
  podLastAppliedToAPIserver *v1.Pod
  podLastAppliedToProvider *v1.Pod
  apiServerPodLastUpdated time.Time
  providerPodLastUpdated time.Time
  lastReconcilationLoopSuccess time.Time
  // … There could be other fields 
}
```

There are a few ideas here:
There is one lock around each pod. The point is that we no longer have 2 (3 if you count deletionQ) workers running
The idea is that the loop logic is idempotent and as long as either apiServerPodLastUpdated or providerPodLastUpdated are later than lastReconcilationLoopSuccess, we need to re-reconcile
There will be a single function responsible for reconciliation that will perform bidirectional reconciliation.
The reconciliation loop can also run periodic retries around reconciliation independent from edge triggered logic (See: The Lease controller has hybrid time based reconciliation, and edge triggered logic when the state of the node changes)

I also suggest we make a small evolution of the PodLifecycleHandler, from:

```go
type PodLifecycleHandler interface {
    // CreatePod takes a Kubernetes Pod and deploys it within the provider.
    CreatePod(ctx context.Context, pod *corev1.Pod) error

    // UpdatePod takes a Kubernetes Pod and updates it within the provider.
    UpdatePod(ctx context.Context, pod *corev1.Pod) error

    // DeletePod takes a Kubernetes Pod and deletes it from the provider.
    DeletePod(ctx context.Context, pod *corev1.Pod) error

    // GetPod retrieves a pod by name from the provider
    GetPod(ctx context.Context, namespace, name string) (*corev1.Pod, error)

    // GetPodStatus retrieves the status of a pod by name from the provider.
    GetPodStatus(ctx context.Context, namespace, name string) (*corev1.PodStatus, error)

    // GetPods retrieves a list of all pods running on the provider
    GetPods(context.Context) ([]*corev1.Pod, error)
}
```

To extend it with:

```go
type PodLifecycleHandlerUIDBased interface {
     // GetPodByUID retrieves a pod by UID from the provider (must be consistent).
    GetPodByUID(ctx context.Context, uid types.UID) (*corev1.Pod, error)

    // GetPodStatus retrieves the status of a pod by name from the provider.
    GetPodStatusByUID(ctx context.Context, uid types.UID) (*corev1.PodStatus, error)
}
```

This will get rid of the icky ambiguity we have in a variety of places in the VK where we can conflate UID of pods with namespace/name because K8s doesn’t enforce namespace/name uniqueness across the lifetime of API Server. I suggest that if the PodeLifecycler handler implements this interface we treat all internal tracking of pods based on UID.

**Comment by hidalgopl** [comment]
@sargun thanks for putting this together :open_hands: I'm not super familiar with VK codebase yet, so sorry if any of those question sounds silly.
If I understand correctly, this would change approach from being reactive to events (and have a logic that tries to figure out what VK should do) to one reconciliation loop where VK will pull pod states from k8s and Provider and based on those decides whether pod should be created/updated/delete, right?
In other words, current approach:
`Add/Update/Delete event -> add pod to Add/Update/Delete queue`
will change to:
`Reconcile -> pull all pod states from VK and Provider -> compare -> decide whether to Add/Update/Delete`
correct ?

This change is fully transparent (no changes in external APIs) for end users (providers using VK), correct?

Regarding extending `PodLifecycleHandler`, migrating path to V1-NG for each provider would be implement those two additional 
`GetPodByUID, GetPodStatusByUID` methods, correct?
Thanks :+1:

**Comment by pires** [comment]
I tend to agree w/ the idempotent approach to the reconciler. This is actually how I'm used to writing controllers, using controller-runtime, which I would suggest to adopt here if it wasn't for the fact that it is non-trivial to get it to watch non-Kubernetes resources, ie the Pod in the Provider.

That said, here's a few comments:

>podReceivedFromAPIServer *v1.Pod

A cached client may or may not receive an event when the Pod changes. The relist period needs to be tuned, if it isn't already so VK doesn't skew much here, or we use an uncached client just for the GET Pod calls.

>podLastAppliedToAPIserver *v1.Pod

We could even inject this in the Pod itself, the same way kubectl does for many resources where it populates the annotation `kubernetes.io/last-applied-configuration` w/ the JSON of the object in that moment. This will allow, for instance, to receive a Pod event, GET Pod and compare the actual Pod w/ what is populated in said annotation, without having to store it in-memory. This will definitely reduce any issues if the VK restarts and loses in-memory representation of what was last applied.

>providerPodLastUpdated 

Maybe we could also populate this as an annotation.

>workNextScheduled *time.Time

Something I really enjoy in controller-runtime is the `Reconciler` function returns `ctrl.Result, error` and `ctrl.Result` is a struct that when success is `nil` but when failure optionally includes the delay. Maybe we can learn from its implementation details?

**Comment by pires** [comment]
Oh, as mentioned over Slack, I don't quite follow this:

>Since the VK knows all pods that are scheduled on it, we can scan over these.

We shouldn't scan and just rely on events which may arrive in near-realtime or during the full-resync period (relist).

**Comment by sargun** [comment]
We should can as a matter of idempotency.

**Comment by pires** [comment]
That's what `full resync period` is all about, the opportunity for controllers to relist the objects they care about and have client-go re-generate events that may have been lost, if any. Kubernetes is after all an eventually-consistent system :)

**Comment by sargun** [comment]
I don't think that's a good idea -- to rely on the resync period. IMHO, ideally these should be "out of phase" with each other. Otherwise, you're combining a lot of load on API server simultaneously.

**Comment by pires** [comment]
You can relist w/ field selectors, namely `nodeName==VKnodename` and reduce the load on both the API server and the client-go internals.

I need you to share your concerns in more detail on why you think full resync is not a good idea. _It is the way_!

**Comment by sargun** [comment]
1. Under the hood, etcd doesn't index this, so the cost is not reduced on the underlying storage
2. Full resync is okay because from my understanding the apiserver is flakey, and you fundamentally need to handle this case. Putting a ton of load on API server from a LIST (iterative read), and THEN potentially doing a bunch of writes seems like a recipe for disaster.

**Comment by pires** [comment]
(1) API server keeps its own cache to avoid overloading etcd when it can.
(2) Unless you can show the problem happening and at what scale it happened, I can't change my opinion. This is how pretty much every single controller out there, built-into Kubernetes and 3rd-party stuff, like Operators ensure they can recover from missing events.
